Common/Arm64Emitter.cpp

```diff
@@ -313,24 +313,40 @@ void ARM64XEmitter::FlushIcache()
 
 void ARM64XEmitter::FlushIcacheSection(u8* start, u8* end)
 {
-	if (cpu_info.sBugs.bExynos8890Invalidation)
-	{
-		// Over invalidate to force this CPU to listen.
-		start = m_startcode + 4096 < start ? start - 4096 : m_startcode;
-		end += 4096;
-	}
-
 #if defined(IOS)
 	// Header file says this is equivalent to: sys_icache_invalidate(start, end - start);
 	sys_cache_control(kCacheFunctionPrepareForExecution, start, end - start);
-#else
-#if (defined(__clang__) && !defined(_M_IX86) && !defined(_M_X64)) || defined(ANDROID)
-	__clear_cache(start, end);
-#else
-#if !defined(_M_IX86) && !defined(_M_X64)
-	__builtin___clear_cache(start, end);
-#endif
-#endif
+#elif !defined(_M_IX86) && !defined(_M_X64)
+	// Code from Dolphin, contributed by the Mono project.
+
+	// Don't rely on GCC's __clear_cache implementation, as it caches
+	// icache/dcache cache line sizes, that can vary between cores on
+	// big.LITTLE architectures.
+	u64 addr, ctr_el0;
+	static size_t icache_line_size = 0xffff, dcache_line_size = 0xffff;
+	size_t isize, dsize;
+
+	__asm__ volatile("mrs %0, ctr_el0" : "=r"(ctr_el0));
+	isize = 4 << ((ctr_el0 >> 0) & 0xf);
+	dsize = 4 << ((ctr_el0 >> 16) & 0xf);
+
+	// use the global minimum cache line size
+	icache_line_size = isize = icache_line_size < isize ? icache_line_size : isize;
+	dcache_line_size = dsize = dcache_line_size < dsize ? dcache_line_size : dsize;
```

Comment:
Hmm, if we're not using the line size we just read, then this can mean a race condition, right?

For example, let's say FlushIcacheSection() is called on BIG cpu, possibly several times.  When it swaps to the LITTLE cpu:
1. If it's before we check the line size, it's fine, we'll get the right one.
2. If it's after we invalidate, then hopefully it's handled correctly.
3. If it's while we're invalidating, we're screwed if it doesn't handle the swap correctly.  We still have the BIG size.

If it's necessary to take the lowest, then we must need to re-read the line size for every iteration of the loop.  It seems to me we shouldn't cache at all?

-[Unknown]